File: core-extra/arazzo-parser/src/main/java/org/evomaster/arazzo/parser/ArazzoParser.java
package org.evomaster.arazzo.parser;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import io.swagger.v3.oas.models.OpenAPI;
import javafx.util.Pair;
import org.evomaster.arazzo.mapper.ArazzoMapper;
import org.evomaster.arazzo.models.domain.ArazzoSpecifications;
import org.evomaster.arazzo.models.dto.ArazzoSpecificationsDTO;
import org.evomaster.arazzo.resolver.ArazzoReferenceResolver;

/**
 * Parse a String containing an Arazzo document into a complete model in ArazzoSpecifications
 */
public class ArazzoParser {

    private static final ObjectMapper JSON_MAPPER = new ObjectMapper().findAndRegisterModules();
    private static final ObjectMapper YAML_MAPPER = new ObjectMapper(new YAMLFactory()).findAndRegisterModules();

    public static ArazzoSpecifications parse(String schemaText, OpenAPI openAPI) {
        Pair<ArazzoSpecificationsDTO, JsonNode> parsed = parseSchemaText(schemaText);
        ArazzoReferenceResolver resolver = new ArazzoReferenceResolver(parsed.getKey().getComponents(), parsed.getValue(), openAPI);
        ArazzoMapper mapper = new ArazzoMapper(resolver);
        return mapper.toDomain(parsed.getKey());
    }

    private static Pair<ArazzoSpecificationsDTO, JsonNode> parseSchemaText(String schemaText) {
        String schemaTextClean = schemaText.replaceAll("^\\s+", "");

        ArazzoSpecificationsDTO arazzoSpecificationsDTO;
        JsonNode arazzoJsonNode;

        try {
            if (schemaTextClean.startsWith("{")) {
                arazzoSpecificationsDTO = JSON_MAPPER.readValue(schemaTextClean, ArazzoSpecificationsDTO.class);
                arazzoJsonNode = JSON_MAPPER.readTree(schemaTextClean);
            } else {
                arazzoSpecificationsDTO = YAML_MAPPER.readValue(schemaTextClean, ArazzoSpecificationsDTO.class);
                arazzoJsonNode = YAML_MAPPER.readTree(schemaTextClean);
            }
        } catch (Exception e) {
            throw new IllegalArgumentException("Problems parsing the Arazzo document", e);
        }

        return new Pair<>(arazzoSpecificationsDTO, arazzoJsonNode);
    }
}

File: core/src/main/kotlin/org/evomaster/core/problem/rest/arazzo/parser/ArazzoParser.kt
package org.evomaster.core.problem.rest.arazzo.parser

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import org.evomaster.core.problem.rest.arazzo.models.ArazzoSpecifications
import com.fasterxml.jackson.module.kotlin.readValue
import org.evomaster.core.problem.rest.arazzo.mapper.ArazzoMapper
import org.evomaster.core.problem.rest.arazzo.models.Workflow
import org.evomaster.core.problem.rest.arazzo.models.raws.ArazzoSpecificationsRaw
import org.evomaster.core.problem.rest.arazzo.resolver.ArazzoReferenceResolver
import org.evomaster.core.problem.rest.schema.SchemaArazzo
import org.evomaster.core.problem.rest.schema.SchemaOpenAPI

object ArazzoParser {

    val jsonMapper = ObjectMapper().findAndRegisterModules()
    val yamlMapper = ObjectMapper(YAMLFactory()).findAndRegisterModules()

    fun parse(schemaText: String, schemaOpenAPI: SchemaOpenAPI) : ArazzoSpecifications {
        val (raw, schemaJsonNode) = parseSchemaText(schemaText)
        val resolver = ArazzoReferenceResolver(raw.components,schemaJsonNode,schemaOpenAPI.schemaParsed)
        val mapper = ArazzoMapper(resolver)
        return mapper.toDomain(raw)
    }

    private fun parseSchemaText(schemaText: String): Pair<ArazzoSpecificationsRaw, JsonNode> {
        val schemaTextClean = schemaText.trimStart()

        var arazzoJsonNode: JsonNode?
        var arazzoSpecificationsRaw: ArazzoSpecificationsRaw?

        try {
            if (schemaTextClean.startsWith("{")) {
                arazzoSpecificationsRaw = jsonMapper.readValue<ArazzoSpecificationsRaw>(schemaTextClean)
                arazzoJsonNode = jsonMapper.readTree(schemaTextClean)
            } else {
                arazzoSpecificationsRaw = yamlMapper.readValue<ArazzoSpecificationsRaw>(schemaTextClean)
                arazzoJsonNode = yamlMapper.readTree(schemaTextClean)
            }
        } catch (e: Exception) {
            throw IllegalArgumentException("Problems parsing the Arazzo document", e)
        }

        return Pair(arazzoSpecificationsRaw, arazzoJsonNode)

    }

}